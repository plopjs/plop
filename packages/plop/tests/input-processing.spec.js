import { resolve, dirname } from "node:path";
import { renderPlop } from "./render.js";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

test("should report a missing plopfile when not copied", async () => {
  // The directory above this repo. We look up towards `plopfile.js` and found one at root otherwise.
  const cwd = resolve(__dirname, "../../../..");
  const { findByError } = await renderPlop([], { cwd });
  expect(await findByError(/\[PLOP\] No plopfile found/)).toBeInTheConsole();
});

test("should show help information on help flag", async () => {
  const { findByText } = await renderPlop(["--help"]);
  const { stdoutArr } = await findByText("Usage:");
  expect(stdoutArr.join("\n")).toMatchSnapshot();
});

test("should show version on version flag", async () => {
  const { findByText } = await renderPlop(["--version"]);
  expect(await findByText(/^[\w\.-]+$/)).toBeInTheConsole();
});

test("should show version on v flag", async () => {
  const { findByText } = await renderPlop(["-v"]);
  expect(await findByText(/^[\w\.-]+$/)).toBeInTheConsole();
});

test("should display inquirer prompts", async () => {
  const { findByText, userEvent } = await renderPlop([], {
    cwd: resolve(__dirname, "./examples/prompt-only"),
  });
  expect(await findByText("What is your name?")).toBeInTheConsole();
  userEvent.keyboard("Joe");
  expect(await findByText("Joe")).toBeInTheConsole();
  userEvent.keyboard("[Enter]");
});

test("Should handle generator prompt", async () => {
  const { findByText, clear, userEvent } = await renderPlop([""], {
    cwd: resolve(__dirname, "./examples/javascript"),
  });

  await findByText("Please choose a generator");

  clear();
  userEvent.keyboard("[ArrowUp]");
  userEvent.keyboard("[ArrowDown]");
  userEvent.keyboard("[Enter]");

  expect(await findByText("this is a test")).toBeInTheConsole();
});

test("Should bypass generator prompt", async () => {
  const { findByText } = await renderPlop(["test"], {
    cwd: resolve(__dirname, "./examples/javascript"),
  });

  expect(await findByText("What is your name?")).toBeInTheConsole();
});

test("Should bypass prompt by input", async () => {
  const { queryByText, findByText } = await renderPlop(["Frank"], {
    cwd: resolve(__dirname, "./examples/prompt-only"),
  });

  expect(await queryByText("What is your name?")).not.toBeInTheConsole();
  expect(
    await findByText("What pizza toppings do you like?")
  ).toBeInTheConsole();
});

test("Should bypass prompt by input placeholder", async () => {
  const { queryByText, findByText, userEvent } = await renderPlop(
    ["_", "Cheese"],
    {
      cwd: resolve(__dirname, "./examples/prompt-only"),
    }
  );

  expect(await findByText("What is your name?")).toBeInTheConsole();
  userEvent.keyboard("[Enter]");
  expect(
    await queryByText("What pizza toppings do you like?")
  ).not.toBeInTheConsole();
});

test("Should bypass prompt by name", async () => {
  const { queryByText, findByText } = await renderPlop(
    ["--", "--name", "Frank"],
    {
      cwd: resolve(__dirname, "./examples/prompt-only"),
    }
  );

  expect(await queryByText("What is your name?")).not.toBeInTheConsole();
  expect(
    await findByText("What pizza toppings do you like?")
  ).toBeInTheConsole();
});

test("Should allow for empty string bypassing", async () => {
  const { queryByText, findByText } = await renderPlop(["--", "--name", `""`], {
    cwd: resolve(__dirname, "./examples/prompt-only"),
  });

  expect(await queryByText("What is your name?")).not.toBeInTheConsole();
  expect(
    await findByText("What pizza toppings do you like?")
  ).toBeInTheConsole();
});

test.todo("Dynamic actions");
