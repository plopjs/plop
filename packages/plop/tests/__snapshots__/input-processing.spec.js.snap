// Vitest Snapshot v1

exports[`should show help information on help flag 1`] = `
"
Usage:
  $ plop                 Select from a list of available generators
  $ plop <name>          Run a generator registered under that name
  $ plop <name> [input]  Run the generator with input data to bypass prompts

Options:
  -h, --help             Show this help display
  -t, --show-type-names  Show type names instead of abbreviations
  -i, --init             Generate a basic plopfile.js
  -v, --version          Print current version
  -f, --force            Run the generator forcefully

 ------------------------------------------------------
  âš   danger waits for those who venture below the line

  --plopfile             Path to the plopfile
  --cwd                  Directory from which relative paths are calculated against while locating the plopfile
  --preload              String or array of modules to require before running plop
  --dest                 Output to this directory instead of the plopfile's parent directory

Examples:
  $ plop
  $ plop component
  $ plop component \\"name of component\\"
"
`;
